### ISSUE TITLE
[CPU] Implement Real Forward Pass in CpuInferenceEngine::forward_parallel

### ISSUE THREAD

**Issue by EffortlessSteven** [EffortlessMetrics/BitNet-rs#292]
# [CPU] Implement Real Forward Pass in `CpuInferenceEngine::forward_parallel`

## Problem Description

The `CpuInferenceEngine::forward_parallel` function in `crates/bitnet-inference/src/cpu.rs` is currently a placeholder implementation that creates dummy tensor results instead of performing actual neural network inference. This stub implementation prevents the CPU inference engine from conducting real BitNet model forward passes with parallel layer processing.

## Environment

- **File**: `crates/bitnet-inference/src/cpu.rs`
- **Function**: `CpuInferenceEngine::forward_parallel` (lines 263-272)
- **Affected Component**: CPU-based BitNet neural network inference
- **Feature Flags**: Requires `--features cpu` for CPU inference functionality

## Current Implementation (Problematic Code)

```rust
/// Forward pass with parallel layer processing
fn forward_parallel(&self, input: &BitNetTensor, _step: usize) -> Result<BitNetTensor> {
    // This is a simplified synchronous version
    // In a full async implementation, we would use model.read().await

    // For now, create a placeholder result
    // In practice, this would require async model access
    let result = BitNetTensor::zeros(&[1, 32000], candle_core::DType::F32, &candle_core::Device::Cpu)?;

    Ok(result)
}
```

## Root Cause Analysis

1. **Stub Implementation**: The function returns hardcoded zero tensors with fixed dimensions `[1, 32000]` instead of processing the actual input tensor through the neural network layers
2. **Missing Model Integration**: No actual model access or forward pass computation is performed
3. **Incomplete Parallelization**: Despite the method name suggesting parallel processing, no Rayon-based parallel layer execution is implemented
4. **Tensor Dimension Mismatch**: Hardcoded output dimensions don't adapt to different model architectures or batch sizes

## Impact Assessment

**Severity**: High - Core functionality completely non-functional
**Affected Users**: All users attempting CPU-based BitNet model inference
**Business Impact**:
- CPU inference engine generates meaningless output
- Token generation produces invalid results
- Cross-validation against reference implementations fails
- Performance benchmarks are misleading

## Reproduction Steps

1. Build with CPU features: `cargo build --no-default-features --features cpu`
2. Load a BitNet model using the CPU inference engine
3. Call `generate_tokens_parallel()` which internally uses `forward_parallel()`
4. **Expected**: Real neural network inference with logits output
5. **Actual**: Zero tensor returned, causing invalid token sampling

### Primary Implementation Approach

Replace the stub with a real forward pass implementation that:

1. **Accesses the actual model**: Use `self.model.blocking_read()` or proper async access patterns
2. **Performs model forward pass**: Call the model's forward method with proper cache handling
3. **Implements parallel layer processing**: Use Rayon for parallel execution across model layers
4. **Handles dynamic tensor shapes**: Adapt output dimensions based on model configuration

### Detailed Implementation Plan

```rust
fn forward_parallel(&self, input: &BitNetTensor, step: usize) -> Result<BitNetTensor> {
    let start_time = std::time::Instant::now();

    // 1. Convert input tensor to appropriate format for model
    let concrete_input = ConcreteTensor::BitNet(input.clone());

    // 2. Access model with proper synchronization
    let model = self.model.blocking_read();

    // 3. Get or create cache for this step
    let mut cache = self.cache.lock().unwrap();

    // 4. Perform forward pass with parallel layer processing
    let output = if self.config.enable_parallel_layers &&
                    model.config().num_layers >= self.config.parallel_layer_threshold {
        // Use parallel layer processing for large models
        self.forward_with_parallel_layers(&model, &concrete_input, &mut *cache, step)?
    } else {
        // Use standard sequential forward pass
        model.forward(&concrete_input, &mut *cache as &mut dyn std::any::Any)?
    };

    // 5. Convert to BitNetTensor and apply final transformations
    let logits = match output {
        ConcreteTensor::BitNet(tensor) => tensor,
        ConcreteTensor::Mock(_) => return Err(InferenceError::InvalidState {
            reason: "Mock tensor returned from real model".to_string()
        }.into()),
    };

    // 6. Update performance metrics
    {
        let mut metrics = self.metrics.lock().unwrap();
        metrics.forward_pass_latency_ms = start_time.elapsed().as_millis() as f64;
    }

    Ok(logits)
}
```

### Additional Methods Required

```rust
/// Parallel layer processing implementation
fn forward_with_parallel_layers(
    &self,
    model: &Box<dyn Model<Config = BitNetConfig>>,
    input: &ConcreteTensor,
    cache: &mut dyn std::any::Any,
    step: usize,
) -> Result<ConcreteTensor> {
    // Implementation using Rayon for parallel layer execution
    // This would require model architecture that supports layer-wise parallelism
    model.forward(input, cache) // Placeholder for now
}
```

### Alternative Approaches Considered

1. **Async-First Implementation**: Use `model.read().await` but requires making the entire inference engine async
2. **Lock-Free Approach**: Use atomic operations or lock-free data structures, but adds complexity
3. **Hybrid Sync/Async**: Current approach with blocking read, maintaining API compatibility

## Implementation Tasks

- [ ] Replace stub implementation with real model forward pass
- [ ] Implement proper model access with `blocking_read()` or async patterns
- [ ] Add cache integration for KV cache management
- [ ] Implement parallel layer processing using Rayon thread pool
- [ ] Add proper error handling for model access and computation failures
- [ ] Update tensor shape handling to be dynamic based on model config
- [ ] Add performance metrics collection for forward pass timing
- [ ] Create comprehensive unit tests for the new implementation
- [ ] Validate cross-compatibility with existing `generate_tokens_parallel()` flow
- [ ] Document the parallel processing approach and performance characteristics

### Unit Tests
```rust
#[test]
fn test_forward_parallel_real_model() {
    // Test with actual BitNet model
    let model = create_test_bitnet_model();
    let engine = CpuInferenceEngine::new(model, backend, config).unwrap();
    let input = BitNetTensor::from_slice(&[1, 2, 3], &[1, 3], &Device::Cpu).unwrap();

    let result = engine.forward_parallel(&input, 0).unwrap();
    assert_ne!(result.shape(), &[1, 32000]); // Should not be hardcoded dimensions
    assert!(!is_all_zeros(&result)); // Should not be zero tensor
}

#[test]
fn test_forward_parallel_performance() {
    // Verify parallel processing improves performance for large models
}

#[test]
fn test_forward_parallel_cache_integration() {
    // Verify KV cache is properly used and updated
}
```

### Integration Tests
- Test with real GGUF model files
- Verify token generation produces meaningful output
- Cross-validate against C++ reference implementation
- Performance benchmarking against sequential implementation

## Success Criteria

- [ ] `forward_parallel()` performs actual neural network computation
- [ ] Output tensor dimensions match model vocabulary size and batch configuration
- [ ] Generated tokens are semantically meaningful (not random/zero)
- [ ] Performance improves with parallel layer processing for large models
- [ ] KV cache integration works correctly across multiple inference steps
- [ ] Cross-validation tests pass with C++ reference implementation
- [ ] Memory usage remains within configured limits
- [ ] All existing tests continue to pass

## Related Issues

This issue is part of a broader effort to eliminate stub implementations across the BitNet.rs codebase:

- **CPU Backend Stubs**: `stub-cpu-backend-tokenize.md`, `stub-cpu-backend-detokenize.md`, `stub-cpu-backend-is-eos-token.md`
- **GPU Inference Stubs**: `stub-gpu-inference-engine-forward-gpu.md`, `stub-gpu-inference-engine-forward-mixed-precision.md`
- **Model Integration**: `stub-model-config.md`, `stub-model-tensors.md`
- **Performance Components**: `stub-generation-stats-placeholders.md`, `stub-kv-cache-update.md`

## Priority

**High Priority** - This is core inference functionality that blocks meaningful neural network computation on CPU devices. Should be implemented as part of the production-ready inference engine initiative.

## Technical Notes

- Requires careful handling of async/sync boundaries in the current architecture
- May need architecture changes if model doesn't support parallel layer processing
- Performance characteristics will depend on model size and CPU core count
- Consider memory pool usage for large tensor allocations
- Ensure thread safety with concurrent cache access

---

**Acceptance Definition**: The CPU inference engine performs real BitNet neural network forward passes, generating meaningful token logits that enable proper autoregressive text generation on CPU devices.
